py/mkrules.cmake

```diff
@@ -0,0 +1,102 @@
+# CMake fragment for MicroPython rules
+
+set(MPY_PY_QSTRDEFS "${MPY_PY_DIR}/qstrdefs.h")
+set(MPY_GENHDR_DIR "${CMAKE_BINARY_DIR}/genhdr")
+set(MPY_MPVERSION "${MPY_GENHDR_DIR}/mpversion.h")
+set(MPY_MODULEDEFS "${MPY_GENHDR_DIR}/moduledefs.h")
+set(MPY_QSTR_DEFS_LAST "${MPY_GENHDR_DIR}/qstr.i.last")
+set(MPY_QSTR_DEFS_SPLIT "${MPY_GENHDR_DIR}/qstr.split")
+set(MPY_QSTR_DEFS_COLLECTED "${MPY_GENHDR_DIR}/qstrdefs.collected.h")
+set(MPY_QSTR_DEFS_PREPROCESSED "${MPY_GENHDR_DIR}/qstrdefs.preprocessed.h")
+set(MPY_QSTR_DEFS_GENERATED "${MPY_GENHDR_DIR}/qstrdefs.generated.h")
+set(MPY_FROZEN_CONTENT "${CMAKE_BINARY_DIR}/frozen_content.c")
+
+target_sources(${MICROPYTHON_TARGET} PRIVATE
+    ${MPY_MPVERSION}
+    ${MPY_QSTR_DEFS_GENERATED}
+    #${MPY_FROZEN_CONTENT}
+)
+
+# Command to force the build of another command
+
+add_custom_command(
+    OUTPUT FORCE_BUILD
+    COMMENT ""
+    COMMAND echo -n
+)
+
+# Generate mpversion.h
+
+add_custom_command(
+    OUTPUT ${MPY_MPVERSION}
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${MPY_GENHDR_DIR}
+    COMMAND python3 ${MPY_DIR}/py/makeversionhdr.py ${MPY_MPVERSION}
+    DEPENDS FORCE_BUILD
+)
+
+# Generate moduledefs.h
+# This is currently hard-coded to support modarray.c only, because makemoduledefs.py doesn't support absolute paths
+
+add_custom_command(
+    OUTPUT ${MPY_MODULEDEFS}
+    COMMAND python3 ${MPY_PY_DIR}/makemoduledefs.py --vpath="." ../../../py/modarray.c > ${MPY_MODULEDEFS}
+    DEPENDS ${MPY_MPVERSION}
+        ${SOURCE_QSTR}
+)
+
+# Generate qstrs
+
+# If any of the dependencies in this rule change then the C-preprocessor step must be run.
+# It only needs to be passed the list of SOURCE_QSTR files that have changed since it was
+# last run, but it looks like it's not possible to specify that with cmake.
+add_custom_command(
+    OUTPUT ${MPY_QSTR_DEFS_LAST}
+    COMMAND ${CMAKE_C_COMPILER} -E \$\(C_INCLUDES\) \$\(C_FLAGS\) \$\(C_DEFINES\) -DNO_QSTR ${SOURCE_QSTR} > ${MPY_GENHDR_DIR}/qstr.i.last
```

Comment:
This was already commented upon, but I wanted to second the necessity for not using generator-specific constructs in the CMake logic.  This breaks non-Makefile generators, and is also fragile for the Makefile-based generators as well since it makes assumptions which may not hold true if the internal implementation details change.  Additionally, it's using global options rather than target-specific compile options.

Reply:
Yes I completely agree.  I think the above-mentioned fix should be a good solution for now.